docs/proposals/templates.md

```diff
@@ -0,0 +1,350 @@
+descriptions of the purpose of each knob
+* Allow application deployers to easily customize exposed values like object names while maintaining referential integrity 
+  between dependent pieces (for example ensuring a pod's labels always match the corresponding selector definition of the service)
+* Support maintaining a library of templates within Kubernetes that can be accessed and instantiated by end users
+* Allow users to quickly and repeatedly deploy instances of well-defined application patterns produced by the community
+* Follow established Kubernetes API patterns by defining new template related APIs which consume+return first class Kubernetes 
+  API (and therefore json conformant) objects.
+
+We do not wish to invent a new turing complete templating language.  There are good options available 
+(e.g. https://github.com/mustache/mustache) for developers who want a completely flexible and powerful solution for creating 
+arbitrarily complex templates with parameters, and tooling can be built around such schemes.
+
+Allowing templates to reference other templates presents versioning+consistency challenges along with making the template 
+no longer a self-contained portable object.  Scenarios necessitating multiple templates can be handled in one of several 
+alternate ways:
+
+* Explicitly constructing a new template that merges the existing templates (tooling can easily be constructed to perform this 
+  operation since the templates are first class api objects).
+* Manually instantiating each template and utilizing [service linking](https://github.com/kubernetes/kubernetes/pull/17543) to share 
+  any necessary configuration data.
+
+This document will also refrain from proposing server APIs or client implementations.  This has been a point of debate, and it makes
+more sense to focus on the template/parameter specification/syntax than to worry about the tooling that will process or manage the
+template objects.  However since there is a desire to at least be able to support a server side implementation, this proposal
+does assume the specification will be k8s API friendly.
+
+## Desired characteristics
+
+* K8S object json-compliant syntax.  This allows server side apis that align with existing k8s apis to be constructed 
+  which consume templates and existing k8s tooling to work with them.  It also allows for api versioning/migration to be managed by
+  the existing k8s codec scheme rather than having to define/introduce a new syntax evolution mechanism.
+  * (Even if they are not part of the k8s core, it would still be good if a server side template processing+managing api supplied 
+    as an ApiGroup consumed the same k8s object schema as the peer k8s apis rather than introducing a new one)
+  * This goal would be met by a solution which wraps some other template content (eg jinja, python, etc) inside a k8s object wrapper
```

Comment:
An opaque template body obstructs other goals, such as customizable names and metadata, referential integrity, and introspectability.